File: user-service/src/main/java/com/gallery/userservice/service/notification/NotificationService.java
package com.gallery.userservice.service.notification;

import org.springframework.stereotype.Service;

import java.util.List;

@Service
public class NotificationService {

    private final List<NotificationStrategy> strategies;

    public NotificationService(List<NotificationStrategy> strategies) {
        this.strategies = strategies;
    }

    public void notifyByEmail(String email, String message) {
        for (NotificationStrategy strategy : strategies) {
            if (strategy instanceof EmailNotification && email != null) {
                strategy.send(email, message);
            }
        }
    }

    public void notifyBySms(String phone, String message) {
        for (NotificationStrategy strategy : strategies) {
            if (strategy instanceof SmsNotification && phone != null) {
                strategy.send(phone, message);
            }
        }
    }
}

File: app/UserService/Services/Notifications/NotificationService.cs
namespace UserService.Services.Notifications
{
    public class NotificationService
    {
        private readonly INotificationStrategy _emailStrategy;
        private readonly INotificationStrategy _smsStrategy;

        public NotificationService(INotificationStrategy emailStrategy, INotificationStrategy smsStrategy)
        {
            _emailStrategy = emailStrategy;
            _smsStrategy = smsStrategy;
        }

        public bool NotifyByEmail(string email, string message)
        {
            return _emailStrategy.Notify(email, message);
        }

        public bool NotifyBySms(string phone, string message)
        {
            return _smsStrategy.Notify(phone, message);
        }
    }
}